### PR TITLE
Grenade grand bug fixing

### PR DESCRIPTION
**Pull request by Macoron**
# Grenade grand bug fixing

Based on rejected PM #2274

### Purpose
New explosion and effect shape system. Major grenade code refactoring. 
Fire tile FX will disappear after explosion (#2553)
Fixed grenade animation with new inventory system (#1755 again)
Fixed bug when destroyed grenade explodes in (0, 0, -100)
Grenade now destroys and respawn correctly (#2231)

### Notes:
This code introduce new class for simple grid iteration (see EffectShape). Right now, there are four type of shapes: square, cross, circle and diamond. Hope this will be useful for magic, spells, etc.

### Please make sure you have followed the self checks below before submitting a PR:

- Code is sufficiently commented
- Code is indented with tabs and not spaces
- The PR does not include any unnecessary .meta, .prefab or <b>.unity (scene) changes</b>
- The PR does not bring up any new compile errors
- The PR has been tested in editor
- Any new files are named using PascalCase (to avoid issues on case sensitive file systems)
- Any new / changed components follow the [Component Development Checklist](https://github.com/unitystation/unitystation/wiki/Component-Development-Checklist)
- Any new objects / items follow the [Creating Items and Objects Guide](https://github.com/unitystation/unitystation/wiki/Creating-Items-and-Objects%3A-Now-With-Prefab-Variants) (especially concerning the use of prefab variants)
- The PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)
- The PR has been tested with round restarts.
- The PR has been tested on moving / rotating / rotated-before-joining matrices (if applicable)

### CODE REVIEW

File: UnityProject/Assets/Scripts/Explosion/Explosion.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+		// Wait until all fire effects are finished
+		yield return WaitFor.Seconds(longestTime);
+
+		Destroy(gameObject);
+	}
+
+	public IEnumerator TimedEffect(Vector3Int position, float time, TileChangeManager tileChangeManager)
+	{
+		position.x -= 1;
+		position.y -= 1;
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/Explosion/Explosion.cs]: If you're compensating for station offsets then it's a wrong way to do it – each matrix has different offsets and it won't work on any other matrices. if you need convert world <> local position then please use `MatrixManager`'s `WorldToLocal`/`LocalToWorld` methods for that

**Review by fomalsd — COMMENTED** [review]
Thanks for making shapes reusable, that simplifies things a lot!
There's one issue that explosion effect doesn't work anywhere except station matrix (try cargo shuttle, for example) – that's because there's some confusion between local and world positions.

File: UnityProject/Assets/Scripts/Explosion/Explosion.cs (hunk 1 above)
Comment (Macoron) [reply UnityProject/Assets/Scripts/Explosion/Explosion.cs]: To be honest, I just copy-pasted this code from previous developer (from Grenade.cs script). No idea what it do though. I will add MatrixManager conversation.

**Comment by Macoron** [comment]
Fixed world/local position error

**Review by fomalsd — APPROVED** [review]
nice, thanks!

**Pull request merged**